sql/schema/v1-core.sql

```diff
@@ -620,77 +607,133 @@ REFERENCING NEW TABLE AS new_table
 FOR EACH STATEMENT
 EXECUTE FUNCTION after_v1_concurrency_slot_insert_function();
 
-CREATE OR REPLACE FUNCTION after_v1_concurrency_slot_delete_function()
-RETURNS trigger AS $$
+CREATE OR REPLACE FUNCTION cleanup_workflow_concurrency_slots(
+    p_strategy_id BIGINT,
+    p_workflow_version_id UUID,
+    p_workflow_run_id UUID
+) RETURNS VOID AS $$
+DECLARE
+    v_sort_id INTEGER;
 BEGIN
-    -- When v1_concurrency_slot is DELETED, we add it to the completed_child_strategy_ids on the parent, but only
-    -- when it's NOT a backoff retry. Backoff retries will continue to consume a workflow concurrency slot.
-    WITH parent_slot AS (
+    -- Get the sort_id for the specific workflow concurrency slot
+    SELECT sort_id INTO v_sort_id
+    FROM v1_workflow_concurrency_slot
+    WHERE strategy_id = p_strategy_id
+      AND workflow_version_id = p_workflow_version_id
+      AND workflow_run_id = p_workflow_run_id;
+
+    -- Acquire an advisory lock for the strategy ID
+    -- There is a small chance of collisions but it's extremely unlikely
+    PERFORM pg_advisory_xact_lock(1000000 * p_strategy_id + v_sort_id);
```

Comment:
This gives us 1 million slots before a collision and still gives us trillions of strategy IDs in the DB. I think this is alright for now. There's no `(bigint, bigint)` signature for advisory locks.